cmd/bosun/expr/elastic.go

```diff
@@ -265,37 +269,47 @@ func (e ElasticHosts) Query(r *ElasticRequest) (*elastic.SearchResult, error) {
 	if err != nil {
 		return nil, err
 	}
-	indicies, err := r.Indexer.Generate(r.Start, r.End)
 	if err != nil {
 		return nil, err
 	}
-	s.Index(indicies...)
-	return s.SearchSource(r.Source).Do()
+	s.Index(r.Indices...)
+	// With IgnoreUnavailable there can be gaps in the indices (i.e. missing days) and we will not error
+	// If no indices match than there will be no successful shards and and error is returned in that case
+	s.IgnoreUnavailable(true)
+	res, err := s.SearchSource(r.Source).Do()
+	if err != nil {
+		return nil, err
+	}
+	if res.Shards == nil {
+		return nil, fmt.Errorf("no shard info in reply, should not be here please file issue")
+	}
+	if res.Shards.Successful == 0 {
+		return nil, fmt.Errorf("no successful shards in result, perhaps the index does exist, total shards: %v, failed shards: %v", res.Shards.Total, res.Shards.Failed)
+	}
+	return res, nil
 }
 
 // ElasticRequest is a container for the information needed to query elasticsearch or a date
 // histogram.
 type ElasticRequest struct {
-	Indexer ESIndexer
+	Indices []string
 	Start   *time.Time
 	End     *time.Time
 	Source  *elastic.SearchSource // This the object that we build queries in
 }
 
 // CacheKey returns the text of the elastic query. That text is the indentifer for
-// the query in the cache
+// the query in the cache. It is a combination of the indices queries and the json query content
 func (r *ElasticRequest) CacheKey() (string, error) {
 	s, err := r.Source.Source()
 	if err != nil {
 		return "", err
 	}
-	var str string
-	var ok bool
-	str, ok = s.(string)
-	if !ok {
-		return "", fmt.Errorf("failed to generate string representation of search source for cache key: %s", s)
+	b, err := json.Marshal(s)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate json representation of search source for cache key: %s", s)
 	}
-	return str, nil
+	return fmt.Sprintf("%v\n%s", r.Indices, b), nil
```

Comment:
additionally, what does the source have that causes it to be a different key? What varies?